metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/sync/BlockSynchronizerProps.scala

```diff
@@ -0,0 +1,196 @@
+
+        prepares = (existingPrepares ++ newPrepares).toList
+        proposerKeys <- Gen.listOfN(prepares.size, Gen.oneOf(federationKeys))
+
+        requests = (prepares zip proposerKeys).zipWithIndex.map {
+          case ((parent, publicKey), idx) =>
+            publicKey -> QuorumCertificate[TestAgreement](
+              phase = Phase.Prepare,
+              viewNumber = ViewNumber(100L + idx),
+              blockHash = parent.id,
+              signature = GroupSignature(())
+            )
+        }
+
+        random <- arbitrary[Int].map(seed => new Random(seed))
+
+      } yield TestFixture(ancestorTree, descendantTree, requests, random)
+    }
+  }
+
+  property("persists") = forAll { (fixture: TestFixture) =>
+    implicit val scheduler = TestScheduler()
+
+    val test = for {
+      fibers <- Task.traverse(fixture.requests) { case (publicKey, qc) =>
+        fixture.synchronizer.sync(publicKey, qc).start
+      }
+      _          <- Task.traverse(fibers)(_.join)
+      downloaded <- fixture.downloadedRef.get
+      persistent <- fixture.persistentRef.get
+      ephemeral  <- fixture.ephemeralRef.get
+    } yield {
+      all(
+        "ephermeral empty" |: ephemeral.isEmpty,
+        "persistent contains all" |: fixture.requests.forall { case (_, qc) =>
+          persistent(Namespace.Blocks).contains(qc.blockHash)
+        },
+        "all uncorrupted" |: persistent(Namespace.Blocks).forall {
+          case (blockHash, block: TestBlock) =>
+            blockHash == block.id && !fixture.isCorrupt(block)
+        },
+        "not download already persisted" |: fixture.ancestorTree.forall {
+          block => !downloaded(block.id)
+        }
+      )
+    }
+
+    // Schedule the execution, using a Future so we can check the value.
+    val testFuture = test.runToFuture
+
+    // Simulate a long time, which should be enough for all downloads to finish.
+    scheduler.tick(1.day)
+
+    testFuture.value.get.get
+  }
+
+  property("no forest") = forAll(
+    for {
+      fixture  <- arbitrary[TestFixture]
+      duration <- Gen.choose(1, fixture.requests.size).map(_ * 500.millis)
+    } yield (fixture, duration)
+  ) { case (fixture: TestFixture, duration: FiniteDuration) =>
+    implicit val scheduler = TestScheduler()
+
+    // Schedule the downloads in the background.
+    Task
+      .traverse(fixture.requests) { case (publicKey, qc) =>
+        fixture.synchronizer.sync(publicKey, qc).startAndForget
+      }
+      .runAsyncAndForget
+
+    // Simulate a some random time, which may or may not be enough to finish the downloads.
+    scheduler.tick(duration)
+
+    // Check now that there the persistent store has just one tree.
+    val test = for {
+      persistent <- fixture.persistentRef.get
+    } yield {
+      persistent(Namespace.Blocks).forall { case (_, block: TestBlock) =>
+        block.parentId.isEmpty ||
+          persistent(Namespace.Blocks).contains(block.parentId)
```

Comment:
This could use comment because it's not obvious that this is a valid condition for a tree, unless one looks at the generator

Reply:
Added :+1: